gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -462,61 +463,82 @@ public int read(ByteBuffer byteBuffer) throws IOException {
+        return bytesRead > 0 ? bytesRead : -1;
+      }
 
-    // The server responds in 2MB chunks, but the client can ask for less than that. We store
-    // the
-    // remainder in bufferedContent and return pieces of that on the next read call (and flush
-    // that buffer if there is a seek).
-    if (bufferedContent != null) {
-      bytesRead += readBufferedContentInto(byteBuffer);
-    }
-    if (!byteBuffer.hasRemaining()) {
-      return bytesRead;
-    }
-    if (positionInGrpcStream == objectSize) {
-      return bytesRead > 0 ? bytesRead : -1;
-    }
+      // read request content overlaps with cached footer data
+      long effectivePosition = positionInGrpcStream + bytesToSkipBeforeReading;
+      if ((footerContent != null) && (effectivePosition >= footerStartOffsetInBytes)) {
+        logger.atFiner().log(
+            "Read request responded with footer content at position '%s'", effectivePosition);
+        bytesRead += readFooterContentIntoBuffer(byteBuffer);
+        return bytesRead;
+      }
 
-    // read request content overlaps with cached footer data
-    long effectivePosition = positionInGrpcStream + bytesToSkipBeforeReading;
-    if ((footerContent != null) && (effectivePosition >= footerStartOffsetInBytes)) {
-      logger.atFiner().log(
-          "Read request responded with footer content at position '%s'", effectivePosition);
-      bytesRead += readFooterContentIntoBuffer(byteBuffer);
-      return bytesRead;
-    }
+      bytesRead += readFromGCS(byteBuffer);
 
-    if (resIterator == null) {
-      OptionalLong bytesToRead = getBytesToRead(byteBuffer);
-      positionInGrpcStream += bytesToSkipBeforeReading;
-      bytesToSkipBeforeReading = 0;
-      requestObjectMedia(bytesToRead);
-      if (bytesToRead.isPresent()) {
-        contentChannelEndOffset = positionInGrpcStream + bytesToRead.getAsLong();
+      if (hasMoreFooterContentToRead(byteBuffer)) {
+        int bytesToWrite = min(byteBuffer.remaining(), footerContent.size());
+        int bytesToSkipInFooter = (int) (positionInGrpcStream - footerStartOffsetInBytes);
+        put(footerContent, bytesToSkipInFooter, bytesToWrite, byteBuffer);
+        positionInGrpcStream += bytesToWrite;
+        bytesRead += bytesToWrite;
       }
-    }
 
-    bytesRead += readObjectContentFromGCS(byteBuffer);
-
-    if (hasMoreFooterContentToRead(byteBuffer)) {
-      int bytesToWrite = min(byteBuffer.remaining(), footerContent.size());
-      int bytesToSkipInFooter = (int) (positionInGrpcStream - footerStartOffsetInBytes);
-      put(footerContent, bytesToSkipInFooter, bytesToWrite, byteBuffer);
-      positionInGrpcStream += bytesToWrite;
-      bytesRead += bytesToWrite;
+      return bytesRead;
+    } catch (InterruptedException e) {
+      cancelCurrentRequest();
+      throw new IOException(e);
     }
+  }
 
-    return bytesRead;
+  private int readFromGCS(ByteBuffer byteBuffer) throws IOException, InterruptedException {
```

Comment:
Footer cache happens to be a ByteString. And retries are simpler there because of context carry over. 

Adding zero-copy deserializer and unifyication of logic *needs* to be done. I did not want to include it in the current PR to avoid complications unless there is a bug in the current impl